src/serde.rs

```diff
@@ -0,0 +1,85 @@
+use serde::de::value::MapDeserializer;
+use serde::de::{Deserialize, Deserializer, Error, IntoDeserializer, MapAccess, Visitor};
+use serde::ser::{Serialize, SerializeMap, Serializer};
+
+use core::fmt::{self, Formatter};
+use core::hash::{BuildHasher, Hash};
+use core::marker::PhantomData;
+
+use crate::ListOrderedMultimap;
+
+impl<K, V, S> Serialize for ListOrderedMultimap<K, V, S>
+where
+    K: Serialize + Hash + Eq,
+    V: Serialize,
+    S: BuildHasher,
+{
+    fn serialize<T>(&self, serializer: T) -> Result<T::Ok, T::Error>
+    where
+        T: Serializer,
+    {
+        let mut map_serializer = serializer.serialize_map(Some(self.values_len()))?;
+        for (key, value) in self {
```

Comment:
Is this the reason you added the `IntoIter` implementation? I think it might be better to do a loop using `pop_front` which will avoid the `Clone` (which presumably the serialization of the key has to do again anyways?)

Reply:
No, it's for the `IntoDeserializer` implementation in line 80. It seems not essential but more an edge case which I haven't understood yet, `IndexMap` implements it as well.  It takes `self` and returns a `MapDeserializer` which in turn requires the `Iterator` trait. I would prefer an iterator item of `(&'iter Key, Value)` which would also be consistent with `Iter` and `IterMut` but the `Iterator` trait doesn't allow items referencing `self`. And the keys cannot be popped out of the `VecList` as they are required for possibly multiple values.

Reply:
Transcoding any-valid-value types is why `IntoDeserializer` was [implemented](https://github.com/bluss/indexmap/issues/86) for `IndexMap`. I think having an owned key-cloning iterator is not necessarily a bad thing, key-cloning is required for an iterator returning owned pairs since `ListOrderedMultimap` reuses one key object for all of this key's values.

Alternatively, `IntoIter` can be made private and maybe in the future the `Iterator` requirement can be lowered backwards-compatibly into `StreamingIterator` which itself implements `Iterator` so that the now private `IntoIter` can eventually become a public `IntoStreamingIter`.

Reply:
OK, I think implementing a cloning `IntoIter` is a bit of a gotcha, but don't really see a workaround here with the `Iterator` limitations. But in the case that this comment is referring to, I still think it's better to do the while loop with `pop_front` as it avoids the cloning here.

Reply:
The `fn serialize<T>(&self, serializer: T)` method (now in line 47) takes self by reference, the loop `for (key, value) in self` iterates over references using `Iter` not `IntoIter`, there is no cloning. Or did you something else?

Reply:
Ah, didn't see it took it by reference, nevermind then.